admin/worker/billing_reporter.go

```diff
@@ -0,0 +1,115 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rilldata/rill/admin/billing"
+	"go.uber.org/zap"
+)
+
+func (w *Worker) reportUsage(ctx context.Context) error {
+	// Get reporting granularity
+	var granularity time.Duration
+	var sqlGrainIdentifier string
+	switch w.admin.Biller.GetReportingGranularity() {
+	case billing.UsageReportingGranularityHour:
+		granularity = time.Hour
+		sqlGrainIdentifier = "hour"
+	case billing.UsageReportingGranularityNone:
+		w.logger.Debug("skipping usage reporting: no reporting granularity configured")
+		return nil
+	default:
+		return fmt.Errorf("unsupported reporting granularity: %s", w.admin.Biller.GetReportingGranularity())
+	}
+
+	t, err := w.admin.DB.FindBillingUsageReportedOn(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get last usage reporting time: %w", err)
+	}
+
+	// start reporting from the last reported time
+	var startTime time.Time
+	if t.IsZero() {
+		startTime = time.Now().UTC().Truncate(granularity).Add(-granularity)
+	} else {
+		startTime = t.UTC()
+	}
+	// report until end of the last hour
+	endTime := time.Now().UTC().Truncate(granularity)
+
+	if !startTime.Before(endTime) {
+		w.logger.Debug("skipping usage reporting: no new usage data available", zap.Time("start_time", startTime), zap.Time("end_time", endTime))
+		return nil
+	}
+
+	// Get metrics client
+	client, ok, err := w.admin.OpenMetricsProject(ctx)
+	if err != nil {
+		w.logger.Error("failed to report usage: unable to get metrics client", zap.Error(err))
+		return err
+	}
+	if !ok {
+		w.logger.Debug("skipping usage reporting: no metrics project configured")
+		return nil
+	}
+
+	u, err := client.GetUsageMetrics(ctx, startTime, endTime, sqlGrainIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get usage metrics: %w", err)
+	}
```

Comment:
Can we have some pagination for this? E.g. if we add other billable metrics, this might get very large and not fit in a single API request (or maybe even in memory).

For example, we could order by event time and paginate using it? And keep paging until we reach a page that is sized below the limit. (And call `UpdateBillingUsageReportedOn` for each page so if it crashes, it won't keep reprocessing the same pages.)